spec/source_map/clip_spec.rb

```diff
@@ -2048,4 +2048,164 @@ def foo
     clip = api_map.clip_at('test.rb', [11, 10])
     expect(clip.infer.to_s).to eq('Bar')
   end
+
+  it 'understands compatible reassignments' do
+    source = Solargraph::Source.load_string(%(
+      class Foo
+        # @return [Foo]
+        def baz; end
+      end
+      bar = Foo.new
+      bar
+      bar = Foo.new
+      bar
+  ), 'test.rb')
+    api_map = Solargraph::ApiMap.new.map(source)
+    clip = api_map.clip_at('test.rb', [6, 6])
+    expect(clip.infer.to_s).to eq('Foo')
+
+    clip = api_map.clip_at('test.rb', [8, 6])
+    expect(clip.infer.to_s).to eq('Foo')
+  end
+
+  it 'understands compatible reassignments from same object' do
+    source = Solargraph::Source.load_string(%(
+      class Foo
+        # @return [Foo]
+        def baz; end
+      end
+      bar = Foo.new
+      bar
+      bar = bar.baz
+      bar
+  ), 'test.rb')
+    api_map = Solargraph::ApiMap.new.map(source)
+    clip = api_map.clip_at('test.rb', [6, 6])
+    expect(clip.infer.to_s).to eq('Foo')
+
+    clip = api_map.clip_at('test.rb', [8, 6])
+    expect(clip.infer.to_s).to eq('Foo')
+  end
+
+  it 'does not get confused by overlapping names between methods and variables' do
+    source = Solargraph::Source.load_string(%(
+      class Foo
+        # @return [Foo]
+        def bar; end
+      end
+      bar = Foo.new
+      bar
+      bar = bar.bar
+      bar
+  ), 'test.rb')
+    api_map = Solargraph::ApiMap.new.map(source)
+    clip = api_map.clip_at('test.rb', [6, 6])
+    expect(clip.infer.to_s).to eq('Foo')
+
+    clip = api_map.clip_at('test.rb', [8, 6])
+    expect(clip.infer.to_s).to eq('Foo')
+  end
```

Comment:
The above are behavior that I managed to break in some of my early attempts, so I figured I'd get them merged in advance to tie my hands in #856 :)